### PR TITLE
Add i18n with a selection of Chinese translations

### PR DESCRIPTION
**Pull request by aeolianeth**
This PR adds initial support for i18n using  https://github.com/lingui/js-lingui/.

A selection of strings have been translated into Chinese (`zh`) for demonstration.

| Language | UI |
| --- | --- |
| zh (chinese) Screen Shot 2021-12-04 at 11 04 02 pm | Screen Shot 2021-12-04 at 10 39 40 pm |
| en Screen Shot 2021-11-27 at 9 04 00 pm | Screen Shot 2021-11-27 at 9 04 09 pm |

## Next steps

1. Mark key strings in the app for translation (using `<trans>` renderless component or `t` function)
2. Add Chinese translations
3. Investigate Crowdin for improved translator experience

### CODE REVIEW

**Comment by aeolianeth** [comment]
cc @mejango @peripheralist @odd-amphora

**Comment by jnoh** [comment]
@aeolianeth would love to see some more translated examples in this PoC:

* How might we handle interpolated strings like [this](https://github.com/jbx-protocol/juice-interface/blob/main/src/components/modals/ConfirmStakeTokensModal.tsx#L77)?
* How do we internationalize numbers?
* How do we internationalize pluralized words?

**Comment by peripheralist** [comment]
sorry it's taken me a while to dig into this, nice poc!

looking briefly at the lingui docs, it seems like we can use an `id` to represent the string element, so going forward we would only have to update text in the locales/**/ directory?

overall this seems like a great pattern. anyone can contribute work translating the site to new or existing locales subfolders, without needing to mess with .tsx

**Comment by aeolianeth** [comment]
@jnoh thanks for those questions!

> How might we handle interpolated strings like this?

LinguiJS handles this out of the box.
https://github.com/jbx-protocol/juice-interface/pull/236/files#diff-80669e3721198a6bddb28097757b85c5559e953c9352fd49f7e46f3818ee1935R19

> How do we internationalize pluralized words?

https://github.com/jbx-protocol/juice-interface/pull/236/files#diff-036d029c69630a6879faf2dc5543920d1b80f6ec29dcb632ff73736ab727d8caR80

> How do we internationalize numbers?
> How do we internationalize datetime?

Ok, this is where LinguiJS might fail. I don't think it supports any number or datetime i18n formatting 😬 . Thanks for calling this out.

`react-intl` [looks like it does](https://formatjs.io/docs/getting-started/installation/#minimal-application), though. I will create a PoC with `react-intl` to see how the dev experience compares.

@peripheralist 

> anyone can contribute work translating the site to new or existing locales subfolders, without needing to mess with .tsx

Exactly 👍 . For now we can accept contributions on GitHub, but in the future we can leverage [collaboration tools](https://translation.io/lingui) that make the translation contributing experience even better.

**Comment by aeolianeth** [comment]
@jnoh actually, I totally forgot but we can already use the `Intl` web api for dates and numbers:

- Numbers: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat
- Dates: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat

For cases where a number is "hard-coded" into the string (i.e. part of the string itself), the translator would translate the entire sentence, including the number.

### On `react-intl`

I had a look into `react-intl` and personally I felt the dev experience of linguijs was nicer (having never used either of them before). They do a good comparison here: https://lingui.js.org/misc/react-intl.html. Not a strong opinion, though.

### Next steps

Before merging I'd like to do the following:
- improve on the implementation slightly
- translate the above-the-fold of the page into Chinese (I'll ask for help in Discord)
- add a translation contribution guide to this repo

If anyone has feedback in the meantime, please comment!

cc @peripheralist

File: src/locales/cn/messages.po [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+msgid "Public goods and services"
+msgstr ""
+
+#: src/components/modals/ConfirmStakeTokensModal.tsx:75
+msgid "Remove {0} ERC20 tokens from your wallet and lock them in the Juicebox protocol."
+msgstr ""
+
+#: src/components/modals/DistributeTokensModal.tsx:80
+msgid "{0, plural, one {# token} other {# tokens}}"
+msgstr ""
```
Comment (aeolianeth) [inline src/locales/cn/messages.po:58]:
```suggestion
msgstr "{0, plural, one {# 令牌}" other {# 令牌}}"
```

File: src/locales/cn/messages.po [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+msgid "Open source businesses"
+msgstr ""
+
+#: src/components/Landing/index.tsx:119
+msgid "Powered by public smart contracts on <0>Ethereum</0>."
+msgstr ""
+
+#: src/components/Landing/index.tsx:54
+msgid "Public goods and services"
+msgstr ""
```
Comment (aeolianeth) [inline src/locales/cn/messages.po:50]:
```suggestion
msgstr "公共产品和服务"
```

File: src/locales/cn/messages.po [hunk 3]
```diff
@@ -0,0 +1,58 @@
+msgid ""
```
Comment (aeolianeth) [inline src/locales/cn/messages.po:1]: Hello translators!

This pull request is our first attempt at internationalizating of Juicebox. We are starting with Chinese translations.

This file is where Chinese translations will live. In this file, each translation has 2 parts:

1. `msgid` - the English string.
1. `msgstr` - the Chinese translation - this is where we need your help!

### How to add translations

For each `msgstr` for this file, add a comment on the line and suggest the Chinese translation. Here is an example: https://github.com/jbx-protocol/juice-interface/pull/236/files#r760041409

File: src/locales/cn/messages.po [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+msgid "Indie artists, devs, creators"
+msgstr ""
+
+#: src/components/Landing/index.tsx:55
+msgid "Open source businesses"
+msgstr ""
+
+#: src/components/Landing/index.tsx:119
+msgid "Powered by public smart contracts on <0>Ethereum</0>."
+msgstr ""
```
Comment (joincc) [inline src/locales/cn/messages.po:46]: 由<0>以太坊</0>公共智能合约提供支持。

File: src/locales/cn/messages.po [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+msgid "Ethereum protocols and DAOs"
+msgstr ""
+
+#: src/components/Landing/index.tsx:52
+msgid "Indie artists, devs, creators"
+msgstr ""
+
+#: src/components/Landing/index.tsx:55
+msgid "Open source businesses"
+msgstr ""
```
Comment (joincc) [inline src/locales/cn/messages.po:42]: msgstr “开源业务”

File: src/locales/cn/messages.po [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+msgid "Design your project"
+msgstr ""
+
+#: src/components/Landing/index.tsx:53
+msgid "Ethereum protocols and DAOs"
+msgstr ""
+
+#: src/components/Landing/index.tsx:52
+msgid "Indie artists, devs, creators"
+msgstr ""
```
Comment (joincc) [inline src/locales/cn/messages.po:38]: msgstr “独立艺术家，开发者，创作者”

File: src/locales/cn/messages.po [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+msgid "Community funding for people and projects"
+msgstr ""
+
+#: src/components/Landing/index.tsx:172
+msgid "Design your project"
+msgstr ""
+
+#: src/components/Landing/index.tsx:53
+msgid "Ethereum protocols and DAOs"
+msgstr ""
```
Comment (joincc) [inline src/locales/cn/messages.po:34]: msgstr “以太坊协议和DAO”

File: src/locales/cn/messages.po [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+msgid "Built for:"
+msgstr ""
+
+#: src/components/Landing/index.tsx:105
+msgid "Community funding for people and projects"
+msgstr ""
+
+#: src/components/Landing/index.tsx:172
+msgid "Design your project"
+msgstr ""
```
Comment (joincc) [inline src/locales/cn/messages.po:30]: msgstr “设计你的项目”

File: src/locales/cn/messages.po [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+msgid "Build a community around a project, fund it, and program its spending. Light enough for a group of friends, powerful enough for a global network of anons."
+msgstr ""
+
+#: src/components/Landing/index.tsx:147
+msgid "Built for:"
+msgstr ""
+
+#: src/components/Landing/index.tsx:105
+msgid "Community funding for people and projects"
+msgstr ""
```
Comment (joincc) [inline src/locales/cn/messages.po:26]: msgstr “为个人和项目提供社区资助”

File: src/locales/cn/messages.po [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+"Project-Id-Version: \n"
+"Report-Msgid-Bugs-To: \n"
+"PO-Revision-Date: \n"
+"Last-Translator: \n"
+"Language-Team: \n"
+"Plural-Forms: \n"
+
+#: src/components/Landing/index.tsx:112
+msgid "Build a community around a project, fund it, and program its spending. Light enough for a group of friends, powerful enough for a global network of anons."
+msgstr ""
```
Comment (joincc) [inline src/locales/cn/messages.po:18]: msgstr “围绕一个项目建立一个社区，为其提供资金，并对其支出进行规划。既能简单易用让一群知己好友轻松创建属于自己的社区项目，也可以强大到能够支持一个全球匿名网络的运作"

File: src/locales/cn/messages.po [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+#: src/components/Landing/index.tsx:147
+msgid "Built for:"
+msgstr ""
```
Comment (joincc) [inline src/locales/cn/messages.po:22]: msgstr "创建的对象:"

File: src/locales/cn/messages.po [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+msgid "Powered by public smart contracts on <0>Ethereum</0>."
+msgstr ""
+
+#: src/components/Landing/index.tsx:54
+msgid "Public goods and services"
+msgstr ""
+
+#: src/components/modals/ConfirmStakeTokensModal.tsx:75
+msgid "Remove {0} ERC20 tokens from your wallet and lock them in the Juicebox protocol."
+msgstr ""
```
Comment (joincc) [inline src/locales/cn/messages.po:54]: msgstr "从你的钱包中去除 {0} ERC20 代币，并把这些代币锁在Juicebox协议中"

File: src/locales/cn/messages.po (hunk 1 above)
Comment (joincc) [inline src/locales/cn/messages.po:58]: msgstr "{0, plural, one {# 代币}" other {# 代币}}"

File: src/locales/cn/messages.po (hunk 4 above)
Comment (aeolianeth) [reply src/locales/cn/messages.po:46]:
```suggestion
msgstr "由<0>以太坊</0>公共智能合约提供支持。"
```

File: src/locales/cn/messages.po (hunk 5 above)
Comment (aeolianeth) [reply src/locales/cn/messages.po:42]:
```suggestion
msgstr "开源业务"
```

File: src/locales/cn/messages.po (hunk 6 above)
Comment (aeolianeth) [reply src/locales/cn/messages.po:38]:
```suggestion
msgstr "独立艺术家，开发者，创作者"
```

File: src/locales/cn/messages.po (hunk 7 above)
Comment (aeolianeth) [reply src/locales/cn/messages.po:34]:
```suggestion
msgstr "以太坊协议和DAO"
```

File: src/locales/cn/messages.po (hunk 8 above)
Comment (aeolianeth) [inline src/locales/cn/messages.po:30]:
```suggestion
msgstr "设计你的项目"
```

File: src/locales/cn/messages.po (hunk 9 above)
Comment (aeolianeth) [reply src/locales/cn/messages.po:26]:
```suggestion
msgstr "为个人和项目提供社区资助"
```

File: src/locales/cn/messages.po (hunk 11 above)
Comment (aeolianeth) [reply src/locales/cn/messages.po:22]:
```suggestion
msgstr "创建的对象:"
```

File: src/locales/cn/messages.po (hunk 10 above)
Comment (aeolianeth) [reply src/locales/cn/messages.po:18]:
```suggestion
msgstr "围绕一个项目建立一个社区，为其提供资金，并对其支出进行规划。既能简单易用让一群知己好友轻松创建属于自己的社区项目，也可以强大到能够支持一个全球匿名网络的运作"
```

File: src/locales/cn/messages.po (hunk 12 above)
Comment (aeolianeth) [reply src/locales/cn/messages.po:54]:
```suggestion
msgstr "从你的钱包中去除 {0} ERC20 代币，并把这些代币锁在Juicebox协议中"
```

File: src/locales/cn/messages.po [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+msgid "Powered by public smart contracts on <0>Ethereum</0>."
+msgstr "由<0>以太坊</0>公共智能合约提供支持。"
+
+#: src/components/Landing/index.tsx:54
+msgid "Public goods and services"
+msgstr "公共产品和服务"
+
+#: src/components/modals/ConfirmStakeTokensModal.tsx:75
+msgid "Remove {0} ERC20 tokens from your wallet and lock them in the Juicebox protocol."
+msgstr "从你的钱包中去除 {0} ERC20 代币，并把这些代币锁在Juicebox协议中"
```
Comment (joincc) [inline src/locales/cn/messages.po:54]: "从你的钱包中扣除 {0} ERC20 代币，并把这些代币锁在Juicebox协议中"

File: src/locales/cn/messages.po [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+msgid "Community funding for people and projects"
+msgstr "为个人和项目提供社区资助"
+
+#: src/components/Landing/index.tsx:172
+msgid "Design your project"
+msgstr "设计你的项目"
+
+#: src/components/Landing/index.tsx:53
+msgid "Ethereum protocols and DAOs"
+msgstr "以太坊协议和DAO"
```
Comment (joincc) [inline src/locales/cn/messages.po:34]: "以太坊协议和 DAO"

File: src/locales/cn/messages.po [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+"Language-Team: \n"
+"Plural-Forms: \n"
+
+#: src/components/Landing/index.tsx:112
+msgid "Build a community around a project, fund it, and program its spending. Light enough for a group of friends, powerful enough for a global network of anons."
+msgstr "围绕一个项目建立一个社区，为其提供资金，并对其支出进行规划。既能简单易用让一群知己好友轻松创建属于自己的社区项目，也可以强大到能够支持一个全球匿名网络的运作"
+
+#: src/components/Landing/index.tsx:147
+msgid "Built for:"
+msgstr "创建的对象:"
```
Comment (joincc) [inline src/locales/cn/messages.po:22]: msgstr "服务的对象:"

File: src/locales/cn/messages.po (hunk 14 above)
Comment (aeolianeth) [reply src/locales/cn/messages.po:34]:
```suggestion
msgstr "以太坊协议和 DAO"
```

File: src/locales/cn/messages.po (hunk 13 above)
Comment (aeolianeth) [reply src/locales/cn/messages.po:54]:
```suggestion
msgstr "从你的钱包中扣除 {0} ERC20 代币，并把这些代币锁在Juicebox协议中"
```

File: src/locales/cn/messages.po (hunk 15 above)
Comment (aeolianeth) [reply src/locales/cn/messages.po:22]:
```suggestion
msgstr "服务的对象:"
```

File: src/i18n.tsx [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+    DEFAULT_LOCALE
+
+  if (!SUPPORTED_LOCALES.includes(locale)) {
+    locale = DEFAULT_LOCALE
+  }
+
+  return locale
+}
+
+async function dynamicActivate(locale: string) {
```
Comment (aeolianeth) [inline src/i18n.tsx:32]: Source: https://lingui.js.org/guides/dynamic-loading-catalogs.html#final-i18n-loader-helper

**Comment by aeolianeth** [comment]
### Update

We've got translations for the the home page (excluding the nav items) \o/

Switch your browser language to Chinese and check it out live on fleek 💃 : https://bafybeifhgpnkfoyteuolga5fqag4qp43e462h3743jt3atrnfdbl4qeaxi.on.fleek.co/#/

Screen Shot 2021-12-04 at 10 39 40 pm

### To-do (before merging this PR)

- Contribution guide - how to contribute to translations

### Next steps

1. Mark strings for translation.
2. Add Chinese translations for >70% of the site.
3. Investigate https://crowdin.com/ for better translator contirbution experience.

cc @peripheralist

**Comment by aeolianeth** [comment]
I've added the contribution guide; this is ready for review!

cc @peripheralist @jnoh

**Comment by peripheralist** [comment]
@aeolianeth great stuff! Thanks for being patient here. Let's just solve this merge conflict and we'll merge it in

**Comment by aeolianeth** [comment]
Thanks @peripheralist! Ready to go 🚀

**Review by peripheralist — APPROVED** [review]

**Pull request merged**